lib/engine/game/g_1858/step/private_closure.rb

```diff
@@ -0,0 +1,111 @@
+
+          def actions(entity)
+            return [] unless entity == current_entity
+            return [] if exchange_corporations(entity).empty?
+
+            %w[choose pass]
+          end
+
+          def description
+            'Exchange Private Companies'
+          end
+
+          def pass_description
+            "Sell #{current_entity.id} to bank"
+          end
+
+          def process_pass(action)
+            @game.close_private(action.entity)
+          end
+
+          def log_skip(_entity); end
+
+          def skip!
+            @game.close_private(current_entity)
+            pass!
+          end
+
+          def choice_available?(_entity)
+            true
+          end
+
+          def choice_name
+            "Exchange #{current_entity.id} for a share"
+          end
+
+          def choices
+            choices = []
+            exchange_corporations(current_entity).each do |corporation|
+              choices << choice_text(corporation, 'treasury') if corporation.num_treasury_shares.positive?
+              choices << choice_text(corporation, 'market') if corporation.num_market_shares.positive?
+            end
+            choices
+          end
+
+          def choice_text(corporation, share_location)
+            "#{corporation.id} #{share_location} share"
+          end
+
+          # Returns a hash with two items:
+          #   corporation: the corporation object
+          #   location: a string, either 'treasury' or 'market'
+          def decode_choice(choice_text)
+            /(?<corp_id>.*) (?<location>.*) share/ =~ choice_text
+            {
+              corporation: @game.corporations.find { |corp| corp.id == corp_id },
+              location: location,
+            }
+          end
+
+          def share_chosen(corporation, share_location)
+            if share_location == 'treasury'
+              corporation.shares.first
+            else
+              @game.share_pool.shares_by_corporation[corporation].first
+            end
+          end
+
+          def process_choose(action)
+            choice = decode_choice(action.choice)
+            corporation = choice[:corporation]
+            share_location = choice[:location]
+            minor = action.entity
+            company = @game.private_company(minor)
+            player = company.owner
+
+            share = share_chosen(corporation, share_location)
+            exchange_for_share(share, corporation, minor, player)
+
+            if share_location == 'treasury'
+              acquire_private(corporation, minor)
+              claim_token(corporation, minor)
+            else
+              company.owner = @game.bank
+              player.companies.delete(company)
+            end
+            @game.close_private(company)
+          end
+
+          def exchange_corporations(minor)
+            @game.corporations.select do |corporation|
+              @game.corporation_private_connected?(corporation, minor) &&
```

Comment:
I see this method relies on the graph, is it possible to call this only when an action is executed via the UI, instead of when determining available actions? Then it could be skipped when `@game.loading` is true so that the engine could do less graph work when loading the game.

I see it is used for some other actions, so that may be something to tackle in a separate PR. (and apologies if this has been discussed elsewhere and I missed it)

Reply:
1. Checking for the corporation having shares first sounds like a good change, anything that can short-circuit graph checks is certainly useful. I wonder if something could be done with auto actions on track lay that look for and cache connections, and then in the private closure step you could just check that cache, but that approach does sound quite a bit more complicated.

2. It sounds like this won't be invoked enough to cause any real issues.

3. I did forget about the fact that normal Track steps already do check `can_lay_tile?` all the time anyway, so no worries there.

Reply:
Scratch what I said for 1, `auto_actions` should be used so that an initial check for connected corporations and privates is done in the browser. It can return `Pass` when there are no connections. Here are a couple examples:

* https://github.com/tobymao/18xx/blob/d2cc06960023d3dc478da11e520086354cd5daf5/lib/engine/game/g_1822/step/minor_acquisition.rb#L34
* https://github.com/tobymao/18xx/blob/08c4329acc894c4094777a5ec296913420d5898f/lib/engine/game/g_18_ny/step/merge_with_nyc.rb#L18

Reply:
OK, I'll look into auto_actions and make this change.

I'll also see if the graph calls in the track and buy/sell shares steps can be removed in the same way, and submit separate pull requests for those.

Reply:
I've made this change, moving the call to `exchange_corporations` out from `actions` into `auto_actions`.